File: 53-maximum-subarray.go
func maxSubArray(nums []int) int {
    if len(nums) == 0{
        return 0
    }
    max := nums[0]
    for _, n := range(nums){
        if n > max{
            max = n
        }
    }
    if max < 0{
        return max
    } 
    
    tmp := 0
    for _, n := range(nums){
        if tmp + n > 0{
            tmp = tmp + n
            if tmp > max{
                max = tmp
            }
        } else {
            tmp = 0
        }
    }
    return max
}

File: bwu/array/53-maximum-subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 1:
            return nums[0]
        
        curr_s, min_s = nums[0], min(0, nums[0])
        res = nums[0]
        
        for i in xrange(1, n):
            curr_s += nums[i]
            res = max(res, curr_s - min_s)
            min_s = min(min_s, curr_s)
            
        return res
